source/common/upstream/health_discovery_service.cc

```diff
@@ -54,12 +89,75 @@ void HdsDelegate::onReceiveInitialMetadata(Http::HeaderMapPtr&& metadata) {
   UNREFERENCED_PARAMETER(metadata);
 }
 
+void HdsDelegate::processMessage(
+    std::unique_ptr<envoy::service::discovery::v2::HealthCheckSpecifier>&& message) {
+  // TODO(lilika): What if the message is empty?
+  for (int i = 0; i < message->health_check_size(); i++) {
+    // Create HdsCluster config
+    ENVOY_LOG(debug, "Creating HdsCluster config");
+    envoy::api::v2::core::BindConfig bind_config;
+
+    clusters_config_.emplace_back();
+    envoy::api::v2::Cluster& cluster_config = clusters_config_.back();
+
+    cluster_config.set_name(message->mutable_health_check(i)->cluster_name());
+    cluster_config.mutable_connect_timeout()->set_seconds(ClusterTimeoutSeconds);
+    cluster_config.mutable_per_connection_buffer_limit_bytes()->set_value(
+        ClusterConnectionBufferLimitBytes);
+
+    // Add endpoints to cluster
+    const int localities_size = message->mutable_health_check(i)->endpoints_size();
+    for (int j = 0; j < localities_size; j++) {
+      auto endpoints = message->mutable_health_check(i)->mutable_endpoints(j);
+      for (int k = 0; k < endpoints->endpoints_size(); k++) {
+        cluster_config.add_hosts()->MergeFrom(*endpoints->mutable_endpoints(k)->mutable_address());
```

Comment:
Same here, you don't want the mutable accessors as your are read only.

Reply:
Also, endpoints can have an optional per-endpoint health check configured. See https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/endpoint/endpoint.proto#L50. Can you add a TODO to deal with this later?